### PR TITLE
NE-1294: Add support for AWS shared VPC in another account

### PR DESCRIPTION
**Pull request by Miciah**
Add support for configuring DNS records in AWS Route 53 using a separate account for the private hosted zone.

* `manifests/00-ingress-credentials-request.yaml`: Add "sts:AssumeRole" to the CredentialsRequest for AWS.
* `pkg/manifests/bindata.go`: Regenerate.
* `pkg/dns/aws/dns.go` (`Config`): Add a `RoleARN` field.
(`NewProvider`): If `config.RoleARN` is set, use it to configure the AWS client using the specified role.
* `pkg/dns/split/dns.go`: New file.  Define a DNS provider implementation that wraps two other DNS providers, using one of them to publish records to the public zone and the other to publish records to the private zone.
(`Provider`): New type.  Store the private and public DNS providers, as well as the private zone so that the `Ensure`, `Delete`, and `Replace` methods can use it to determine whether they are publishing to the public zone or to the private zone.
(`NewProvider`): New function.  Return a split DNS provider.
(`Ensure`, `Delete`, `Replace`): New methods.  Implement the `dns.Provider` interface by calling the respective methods on the wrapped private and public DNS providers.
* `pkg/dns/split/dns_test.go` (`TestSplitDNSProvider`): Verify that the split DNS provider correctly dispatches to the private or public DNS provider as appropriate, using `fakeProvider`.
(`fakeProvider`): New type.  Define a fake named DNS provider that records its name when invoked.
(`Ensure`, `Delete`, `Replace`): New methods for `fakeProvider` to record invocations and implement the `dns.Provider` interface.
(`newFakeProvider`): New function.  Return a fake provider.
* `pkg/operator/controller/dns/controller.go` (`Config`): Add a `PrivateHostedZoneAWSEnabled` field to indicate whether the "PrivateHostedZoneAWS" feature gate is enabled.
(`createDNSProvider`): Use the new split DNS provider and the AWS DNS provider's new `RoleARN` configuration option to configure separate DNS providers for public and private zones when a role ARN for the private zone is specified in the cluster infrastructure config if the "PrivateHostedZoneAWS" feature gate is enabled.
* `pkg/operator/operator.go` (`New`): Check the "PrivateHostedZoneAWS" feature gate and specify it in the DNS controller config.

---

This PR is a draft because the `Infrastructure.status.platformStatus.aws.privateHostedZoneRole` API field doesn't exist yet and this is completely untested.

### CODE REVIEW

**Comment by Miciah** [comment]
Rebased in https://github.com/openshift/cluster-ingress-operator/compare/e336dd1c85c9878d08f0f5889505eb364f14a810..584d8fde704fda9b92ad8b91479833e3b47a5f30.

**Comment by Miciah** [comment]
https://github.com/openshift/cluster-ingress-operator/compare/584d8fde704fda9b92ad8b91479833e3b47a5f30..064d8477b16e8551802a8fb3025bfe65eca5c8e2 adds a check for the "SharedVPC" feature gate.

**Comment by gcs278** [comment]
I'm still keeping an eye on this, thanks for the responsiveness, updates look good to me

**Comment by Miciah** [comment]
https://github.com/openshift/cluster-ingress-operator/compare/c064b43a8dafbb43d5e240e3a875ddfa4d85e857..1bee17560eed0fd8ffdd9876f4b5a35344c2ee1b bumps to https://github.com/openshift/api/pull/1460/commits/55be8ed358e5557f9ea62e364ee6a088693c3b16, which makes `Spec.Platform` a non-pointer value again.

**Comment by Miciah** [comment]
https://github.com/openshift/cluster-ingress-operator/compare/1bee17560eed0fd8ffdd9876f4b5a35344c2ee1b..112d591d6c8dcbc36a8f98993cc9a9387acb071b rebases to get #936 and replaces the https://github.com/openshift/api/commit/55be8ed358e5557f9ea62e364ee6a088693c3b16 vendor bump with the actual merge commit, https://github.com/openshift/api/commit/83abb00e268469ba4c898c2995ee7f882a35e8a6.

**Comment by gcs278** [comment]
hypershift issues:
```
```

File: go.mod [hunk 1]
```diff
@@ -18,7 +18,7 @@ require (
 	github.com/google/go-cmp v0.5.9
 	github.com/jongio/azidext/go/azidext v0.4.0
 	github.com/kevinburke/go-bindata v3.11.0+incompatible
-	github.com/maistra/istio-operator v0.0.0-20230322122339-793794762e67
+	github.com/maistra/istio-operator v0.0.0-20230530131601-94a89812be1c
```
[the remark this answers is not in the document]
Comment (Miciah) [reply go.mod:21]: The Go tooling did this; I'm not sure why.  I'll drop it.

[the remark this answers is not in the document]
Comment (Miciah) [reply go.mod]: https://github.com/openshift/cluster-ingress-operator/compare/112d591d6c8dcbc36a8f98993cc9a9387acb071b..ae7a8d24b2080c8593f351fb26d9dfe766fa6d10 drops the gratuitous bump.

**Comment by gcs278** [comment]
Known issue `TestAWSELBConnectionIdleTimeout`, will probably need to be overridden.
/test e2e-aws-operator
Same issue I've been seeing with e2e-hypershift:
```
ON   AVAILABLE   PROGRESSING   SINCE   STATUS
version             False       True          21m     Unable to apply 4.13.0-0.nightly-2023-01-17-084735: some cluster operators are not available
error: timed out waiting for the condition on clusterversions/version
2023-05-31 14:09:21+00:00 Clusteroperators not yet ready
```
/test e2e-hypershift

**Comment by gcs278** [comment]
e2e-aws-operator: must gather failure
/test e2e-aws-operator
same `e2e-hypershift-create-management-cluster container test` failure
/test e2e-hypershift

**Comment by gcs278** [comment]
`TestHeaderNameCaseAdjustment` and must gather failure
/test e2e-aws-operator

**Comment by gcs278** [comment]
`TestAWSELBConnectionIdleTimeout` known issue.
/test e2e-aws-operator

**Comment by gcs278** [comment]
must gather failures
/test e2e-aws-operator

**Comment by gcs278** [comment]
CI looks good now.
`e2e-azure-ovn` optional test failed with `[sig-instrumentation] Prometheus [apigroup:image.openshift.io] when installed on the cluster should report telemetry [Late] [Skipped:Disconnected] [Suite:openshift/conformance/parallel`
Seems unrelated.

**Comment by Miciah** [comment]
https://github.com/openshift/cluster-ingress-operator/compare/ae7a8d24b2080c8593f351fb26d9dfe766fa6d10..5a965ff8f7b0612025ea3c4748723e94f0f92cd0 rebases for #939.

**Comment by gcs278** [comment]
`TestAWSELBConnectionIdleTimeout` known failure
/test e2e-aws-operator

**Comment by gcs278** [comment]
https://github.com/openshift/api/pull/1480 just merged. I think we will probably need to rebase if the validation update needs to be revendoring; however I am not sure it matters for this PR. It may only be a CRD-level validation.

**Comment by Miciah** [comment]
> [openshift/api#1480](https://github.com/openshift/api/pull/1480) just merged. I think we will probably need to rebase if the validation update needs to be revendoring; however I am not sure it matters for this PR. It may only be a CRD-level validation.

I don't think it matters, but I went ahead and rebased anyway.

**Comment by gcs278** [comment]
Thanks - better safe than sorry.
/lgtm
/approve

**Comment by lihongan** [comment]
/label qe-approved

With Installer QE's helping, we got a cluster and tested the PR, it looks good.
```
$ oc get co/ingress
NAME      VERSION                                                AVAILABLE   PROGRESSING   DEGRADED   SINCE   MESSAGE
ingress   4.14.0-0.test-2023-06-05-024830-ci-ln-xf681l2-latest   True        False         False      19h     

$ oc get dns.config cluster -oyaml | yq .spec
baseDomain: yunjiang-xvpc1.qe.devcluster.openshift.com
platform:
  aws:
    privateZoneIAMRole: arn:aws:iam::123456:role/yunjiang-rol1
  type: AWS
privateZone:
  id: XYZZZZ
publicZone:
  id: ABC

### creating more ingresscontrollers and all work
$ oc -n openshift-ingress-operator get ingresscontroller
NAME      AGE
default   19h
extlb     6h16m
lb-int    50m

```
note: the test results is from a Passthrough mode cluster. We're still seeing some issues on STS mode cluster but looks like configuration issue and will not block this PR to be merged.

**Comment by Miciah** [comment]
Rebased for #906.

**Comment by Miciah** [comment]
https://github.com/openshift/cluster-ingress-operator/compare/dbf16265f1718bc51476d36e75e709bdd8991b00..7251aaa9a30f3759c2731d061207d2f63cd945de adds `sts:AssumeRole` to the AWS CredentialsRequest.

**Comment by gcs278** [comment]
`sts:AssumeRole` update looks good to me - just based on the [slack thread](https://redhat-internal.slack.com/archives/C054PJX8LE8/p1686037441441749)
/lgtm
/approved

**Comment by gcs278** [comment]
`TestAWSELBConnectionIdleTimeout` known failure
/test e2e-aws-operator
`TestInternalLoadBalancer` known failure
/test e2e-gcp-operator
Same test-node-pool failure I've been seeing:
/test e2e-hypershift

**Comment by gcs278** [comment]
`Prometheus [apigroup:image.openshift.io] when installed on the cluster should report telemetry`
/test e2e-gcp-ovn

**Comment by Miciah** [comment]
[e2e-aws-operator failed](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-ingress-operator/928/pull-ci-openshift-cluster-ingress-operator-master-e2e-aws-operator/1666531402968993792) on `TestAWSELBConnectionIdleTimeout`, which is a known issue ([OCPBUGS-13810](https://issues.redhat.com/browse/OCPBUGS-13810)) with a medium flake-rate ([search.ci reports](https://search.ci.openshift.org/?search=FAIL%3A+TestAll%2Fparallel%2FTestAWSELBConnectionIdleTimeout&maxAge=24h&context=1&type=all&name=cluster-ingress-operator&excludeName=&maxMatches=5&maxBytes=20971520&groupBy=job) a 44% impact over the past 24 hours as of this writing).
/test e2e-aws-operator

[e2e-gcp-operator failed](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-ingress-operator/928/pull-ci-openshift-cluster-ingress-operator-master-e2e-gcp-operator/1666531402985771008) because `TestUnmanagedDNSToManagedDNSInternalIngressController` and `TestScopeChange` failed, which is a known issue ([OCPBUGS-13106](https://issues.redhat.com/browse/OCPBUGS-13106)) with a medium flake-rate ([search.ci reports](https://search.ci.openshift.org/?search=FAIL%3A+TestAll%2Fparallel%2F%28TestUnmanagedDNSToManagedDNSInternalIngressController%7CTestScopeChange%29&maxAge=24h&context=1&type=all&name=cluster-ingress-operator&excludeName=&maxMatches=5&maxBytes=20971520&groupBy=job) a 50% impact over the past 24 hours as of this writing).
/test e2e-gcp-operator

[e2e-hypershift failed](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-ingress-operator/928/pull-ci-openshift-cluster-ingress-operator-master-e2e-hypershift/1666531403015131136) because `TestNodePool/NodePool_Tests_Group/TestNodePoolReplaceUpgrade` failed.  We don't know how to diagnose e2e-hypershift failures, but we have seen a lot of flakiness in general with this CI job.
/test e2e-hypershift

**Comment by Miciah** [comment]
Regarding the e2e-hypershift failure, [search.ci reports](https://search.ci.openshift.org/?search=FAIL%3A+TestNodePool%2FNodePool_Tests_Group%2FTestNodePoolReplaceUpgrade&maxAge=24h&context=1&type=bug%2Bissue%2Bjunit&name=&excludeName=&maxMatches=5&maxBytes=20971520&groupBy=job) rather a lot of failures for `TestNodePool/NodePool_Tests_Group/TestNodePoolReplaceUpgrade` over the past 24 hours:

> pull-ci-openshift-cluster-ingress-operator-master-e2e-hypershift (all) - 10 runs, 80% failed, 38% of failures match = 30% impact

> pull-ci-openshift-hypershift-main-e2e-aws (all) - 20 runs, 25% failed, 20% of failures match = 5% impact

> pull-ci-openshift-machine-config-operator-master-e2e-hypershift (all) - 12 runs, 83% failed, 10% of failures match = 8% impact

> pull-ci-openshift-cluster-network-operator-master-e2e-hypershift-ovn (all) - 3 runs, 100% failed, 67% of failures match = 67% impact

**Comment by Miciah** [comment]
[e2e-hypershift failed](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-ingress-operator/928/pull-ci-openshift-cluster-ingress-operator-master-e2e-hypershift/1666569867236478976) on `TestNodePool/NodePool_Tests_Group/TestNodePoolReplaceUpgrade` again.
@enxebre, can you help diagnose these failures?
/test e2e-hypershift

**Comment by Miciah** [comment]
e2e-hypershift also failed on #927 with the same test failure, and e2e-hypershift had previously passed on #927.

**Comment by Miciah** [comment]
/test e2e-hypershift
now that https://github.com/openshift/hypershift/pull/2665 has merged.

**Pull request merged**